builders/sol/server.go

```diff
@@ -0,0 +1,176 @@
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package main
+
+import (
+	"archive/tar"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+)
+
+// TODO: HERE: improving error messages
+// TODO: HERE: normalize with zip output?
+
+func runServer() {
+	addr := ":2222"
+	fmt.Println("Listening at", addr)
+	http.HandleFunc("/build/", handleBuild)
+	http.HandleFunc("/list", handleList)
+	log.Fatal(http.ListenAndServe(addr, nil))
+}
+
+func listPlatforms() []string {
+	dir, _ := os.Open(initialDir)
+	fis, _ := dir.Readdir(0)
+
+	var platforms []string
+	for _, fi := range fis {
+		if fi.IsDir() && strings.HasPrefix(fi.Name(), "platform-") {
+			platforms = append(platforms, fi.Name())
+		}
+	}
+
+	dir.Close()
+	return platforms
+}
+
+func handleList(w http.ResponseWriter, r *http.Request) {
+	for _, p := range listPlatforms() {
+		fmt.Fprintln(w, p)
+	}
+}
+
+func platformExists(platform string) bool {
+	for _, p := range listPlatforms() {
+		if p == platform {
+			return true
+		}
+	}
+	return false
+}
+
+func clientError(w http.ResponseWriter, format string, a ...interface{}) {
+	fmt.Print("Request failed: ")
+	fmt.Printf(format, a...)
+	fmt.Println()
+	fmt.Fprintf(w, format, a...)
+	fmt.Fprintln(w)
+}
+
+func internalError(w http.ResponseWriter, format string, a ...interface{}) {
+	fmt.Print("Internal error: ")
+	fmt.Printf(format, a...)
+	fmt.Println()
+	fmt.Fprintln(w, "Internal error")
+}
+
+func compile(dir string, platform string) (*bytes.Buffer, error) {
+	var out bytes.Buffer
+	mw := io.MultiWriter(&out, os.Stdout)
+
+	// TODO: check permission denied error
+
+	cmd := exec.Command(initialDir + "/" + platform + "/compile")
```

Comment:
Is there any way the environment can be controlled by whoever sends the request? Or, better, yet, will `exec.Command()` give the command a completely new, empty environment, in order to avoid possible contamination?